File: src/Main/150/146.cc
#include <unordered_map>

using namespace std;

class Node {
public:
  int val;
  Node *next = nullptr;
  Node *prev = nullptr;
  Node(int val) : val(val) {}
};

class LRUCache {
  unordered_map<int, pair<int, Node *>> mp;
  Node *head = nullptr;
  Node *tail = nullptr;
  int size = 0;
  int capacity;

public:
  LRUCache(int capacity) : capacity(capacity) {}

  void move(Node *p) {
    if (p == tail) {
      return;
    }
    if (p == head) {
      head = head->next;
      if (head) {
        head->prev = nullptr;
      }
    } else {
      p->prev->next = p->next;
      if (p->next) {
        p->next->prev = p->prev;
      }
    }
    tail->next = p;
    p->prev = tail;
    p->next = nullptr;
    tail = p;
  }

  int get(int key) {
    if (mp.find(key) != mp.end()) {
      int val = mp[key].first;
      Node *p = mp[key].second;
      move(p);
      return val;
    }
    return -1;
  }

  void put(int key, int value) {
    if (mp.find(key) != mp.end()) {
      Node *p = mp[key].second;
      move(p);
      mp[key] = make_pair(value, p);
      return;
    }

    Node *v = new Node(key);
    if (size == 0) {
      head = v;
      tail = v;
    } else {
      tail->next = v;
      v->prev = tail;
      tail = v;
    }
    mp[key] = make_pair(value, v);
    size++;

    while (size > capacity) {
      int k = head->val;
      Node *p = head->next;
      if (p) {
        p->prev = nullptr;
      }
      delete head;
      head = p;
      size--;
      mp.erase(k);
    }
  }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: src/archive/Zon/AmazonPrepOld/LinkedList/146.go
package LinkedList

type DLLNode[T any] struct {
	val  T
	key  T
	next *DLLNode[T]
	prev *DLLNode[T]
}

// the idea is maintain a linked list, and pop left node once the cache is full
type LRUCache struct {
	head     *DLLNode[int]
	tail     *DLLNode[int]
	mp       map[int]*DLLNode[int]
	capacity int
}

func Constructor(capacity int) LRUCache {
	return LRUCache{
		nil,
		nil,
		make(map[int]*DLLNode[int]),
		capacity,
	}
}

func (this *LRUCache) removeNode(node *DLLNode[int]) {
	if this.head == node && this.tail == node {
		this.head = nil
		this.tail = nil
		return
	}

	if this.head == node {
		this.head = node.next
		this.head.prev = nil
	} else if this.tail == node {
		this.tail = node.prev
		this.tail.next = nil
	} else {
		node.prev.next = node.next
		node.next.prev = node.prev
	}

	node.prev = nil
	node.next = nil
}

func (this *LRUCache) insertNode(node *DLLNode[int]) {
	node.next = nil
	node.prev = nil
	if this.tail == nil {
		// this is the first node
		this.head = node
		this.tail = node
	} else {
		// just add to the end
		this.tail.next = node
		node.prev = this.tail
		this.tail = node
	}
}

// lru -> if we get some key, need to move it to the back
func (this *LRUCache) Get(key int) int {
	if value, ok := this.mp[key]; ok {
		this.removeNode(value)
		this.insertNode(value)
		return value.val
	}

	return -1
}

func (this *LRUCache) Put(key int, value int) {
	if node, ok := this.mp[key]; ok {
		node.val = value
		this.removeNode(node)
		this.insertNode(node)
		return
	}

	if len(this.mp) == this.capacity {
		evictKey := this.head.key
		this.removeNode(this.head)
		delete(this.mp, evictKey)
	}

	newNode := &DLLNode[int]{key: key, val: value}
	this.insertNode(newNode)
	this.mp[key] = newNode
}
